### PR TITLE
Support nested column projection

### LINKED ISSUE #204: nested projection problem
```python
In [1]: uri = "/home/chang/Downloads/datasets/coco/coco.lance"

In [2]: import lance

In [3]: ds = lance.dataset(uri)

In [4]: ds.schema
Out[4]: 
license: int64
file_name: string
coco_url: extension<image[uri]<ImageUriType>>
height: int64
width: int64
date_captured: timestamp[ns]
flickr_url: extension<image[uri]<ImageUriType>>
image_id: int64
split: dictionary<values=string, indices=int8, ordered=0>
image_uri: extension<image[uri]<ImageUriType>>
annotations: struct<segmentation: list<item: struct<counts: list<item: int32>, polygon: list<item: list<item: float>>, size: list<item: int32>>>, area: list<item: double>, iscrowd: list<item: bool>, bbox: list<item: fixed_size_list<item: float>[4]>, category_id: list<item: int16>, id: list<item: int64>, supercategory: list<item: dictionary<values=string, indices=int8, ordered=0>>, name: list<item: dictionary<values=string, indices=int8, ordered=0>>>
  child 0, segmentation: list<item: struct<counts: list<item: int32>, polygon: list<item: list<item: float>>, size: list<item: int32>>>
      child 0, item: struct<counts: list<item: int32>, polygon: list<item: list<item: float>>, size: list<item: int32>>
          child 0, counts: list<item: int32>
              child 0, item: int32
          child 1, polygon: list<item: list<item: float>>
              child 0, item: list<item: float>
                  child 0, item: float
          child 2, size: list<item: int32>
              child 0, item: int32
  child 1, area: list<item: double>
      child 0, item: double
  child 2, iscrowd: list<item: bool>
      child 0, item: bool
  child 3, bbox: list<item: fixed_size_list<item: float>[4]>
      child 0, item: fixed_size_list<item: float>[4]
          child 0, item: float
  child 4, category_id: list<item: int16>
      child 0, item: int16
  child 5, id: list<item: int64>
      child 0, item: int64
  child 6, supercategory: list<item: dictionary<values=string, indices=int8, ordered=0>>
      child 0, item: dictionary<values=string, indices=int8, ordered=0>
  child 7, name: list<item: dictionary<values=string, indices=int8, ordered=0>>
      child 0, item: dictionary<values=string, indices=int8, ordered=0>
image: extension<image[binary]<ImageBinaryType>>

In [5]: ds.scanner(columns=["annotations.name"]).to_table()
---------------------------------------------------------------------------
ArrowTypeError                            Traceback (most recent call last)
Cell In [5], line 1
----> 1 ds.scanner(columns=["annotations.name"]).to_table()

File ~/.venv/lance/lib/python3.10/site-packages/pyarrow/_dataset.pyx:2577, in pyarrow._dataset.Scanner.to_table()

File ~/.venv/lance/lib/python3.10/site-packages/pyarrow/error.pxi:144, in pyarrow.lib.pyarrow_internal_check_status()

File ~/.venv/lance/lib/python3.10/site-packages/pyarrow/error.pxi:123, in pyarrow.lib.check_status()

ArrowTypeError: struct fields don't match or are in the wrong order: Input fields: struct<name: list<item: dictionary<values=string, indices=int8, ordered=0>>> output fields: struct<segmentation: list<item: struct<counts: list<item: int32>, polygon: list<item: list<item: float>>, size: list<item: int32>>>, area: list<item: double>, iscrowd: list<item: bool>, bbox: list<item: fixed_size_list<item: float>[4]>, category_id: list<item: int16>, id: list<item: int64>, supercategory: list<item: dictionary<values=string, indices=int8, ordered=0>>, name: list<item: dictionary<values=string, indices=int8, ordered=0>>>
```

### PR DESCRIPTION
**Pull request by eddyxu**
Closes #204

### CODE REVIEW

File: cpp/src/lance/format/schema.cc [hunk 1]
```diff
@@ -361,7 +361,7 @@ std::shared_ptr<Field> Field::Copy(bool include_children) const {
 }
 
 std::shared_ptr<Field> Field::Project(const std::shared_ptr<::arrow::Field>& arrow_field) const {
-  assert(name_ == arrow_field->name());
+  //  assert(name_ == arrow_field->name());
```
Comment (changhiskhan) [inline cpp/src/lance/format/schema.cc:364]: Why remove this?

File: cpp/src/lance/arrow/scanner_test.cc [hunk 2]
```diff
@@ -216,7 +216,8 @@ TEST_CASE("Filter over empty list") {
 {"ints": 1, "floats": [0.1, 0.2]},
 {"ints": 2},
 {"ints": 3, "floats": [11.1]}
-])").ValueOrDie();
+])")
+               .ValueOrDie();
```
Comment (changhiskhan) [inline cpp/src/lance/arrow/scanner_test.cc:220]: Indent?

File: cpp/src/lance/arrow/scanner_test.cc [hunk 3]
```diff
@@ -239,7 +240,8 @@ TEST_CASE("Filter with limit") {
   auto t = TableFromJSON(schema, R"([
 {"ints": 1, "floats": [0.1, 0.2]},
 {"ints": 2, "floats": [11.1]}
-])").ValueOrDie();
+])")
+               .ValueOrDie();
```
Comment (changhiskhan) [inline cpp/src/lance/arrow/scanner_test.cc:244]: Same

**Review by changhiskhan — APPROVED** [review]

File: cpp/src/lance/format/schema.cc (hunk 1 above)
Comment (eddyxu) [reply cpp/src/lance/format/schema.cc:364]: oh, in a debug change, lemme revert this.

File: cpp/src/lance/arrow/scanner_test.cc (hunk 2 above)
Comment (eddyxu) [reply cpp/src/lance/arrow/scanner_test.cc:220]: Weird , Clion auto formats into this.

File: cpp/src/lance/arrow/scanner_test.cc (hunk 3 above)
Comment (eddyxu) [reply cpp/src/lance/arrow/scanner_test.cc]: fixed.

**Pull request merged**